kolibri/core/content/utils/import_export_content.py

```diff
@@ -170,52 +210,160 @@ def get_import_export_data(  # noqa: C901
+        ).values("id", "file_size", "extension")
+        if available is not None:
+            file_objects = file_objects.filter(available=available)
+        for f in file_objects:
+            queried_file_objects[f["id"]] = f
+
+        if topic_thumbnails:
+            # Do a query to get all the descendant and ancestor topics for this segment
+            segment_boundaries = nodes_query.aggregate(
+                min_boundary=Min("lft"), max_boundary=Max("rght")
+            )
+            segment_topics = ContentNode.objects.filter(
+                channel_id=channel_id, kind=content_kinds.TOPIC
+            ).filter(
+                Q(
+                    lft__lte=segment_boundaries["min_boundary"],
+                    rght__gte=segment_boundaries["max_boundary"],
+                )
+                | Q(
+                    lft__lte=segment_boundaries["max_boundary"],
+                    rght__gte=segment_boundaries["min_boundary"],
+                )
+            )
+
             file_objects = LocalFile.objects.filter(
-                files__contentnode__in=nodes_segment
+                files__contentnode__in=segment_topics,
             ).values("id", "file_size", "extension")
             if available is not None:
                 file_objects = file_objects.filter(available=available)
             for f in file_objects:
                 queried_file_objects[f["id"]] = f
 
-            if topic_thumbnails:
-                # Do a query to get all the descendant and ancestor topics for this segment
-                segment_boundaries = nodes_segment.aggregate(
-                    min_boundary=Min("lft"), max_boundary=Max("rght")
-                )
-                segment_topics = ContentNode.objects.filter(
-                    channel_id=channel_id, kind=content_kinds.TOPIC
-                ).filter(
-                    Q(
-                        lft__lte=segment_boundaries["min_boundary"],
-                        rght__gte=segment_boundaries["max_boundary"],
-                    )
-                    | Q(
-                        lft__lte=segment_boundaries["max_boundary"],
-                        rght__gte=segment_boundaries["min_boundary"],
-                    )
-                )
-
-                file_objects = LocalFile.objects.filter(
-                    files__contentnode__in=segment_topics,
-                ).values("id", "file_size", "extension")
-                if available is not None:
-                    file_objects = file_objects.filter(available=available)
-                for f in file_objects:
-                    queried_file_objects[f["id"]] = f
-
-        min_boundary += dynamic_chunksize
-
     files_to_download = list(queried_file_objects.values())
 
     total_bytes_to_transfer = sum(map(lambda x: x["file_size"] or 0, files_to_download))
     return number_of_resources, files_to_download, total_bytes_to_transfer
 
 
+def get_content_nodes_selectors(channel_id, nodes_queries_list):
```

Comment:
I feel like there's a more efficient way to do this with a hierarchical aggregation, similar to how we do our annotation logic. I will try to sketch out what that might look like and get back to you here.

Reply:
Yeah, I think you're right that there's probably a better way. I opted for the simple to read, procedural approach since it's a one-off task that users already understand will take a length of time measured in minutes (so a few seconds generating a manifest isn't a big deal as long as we aren't pegging their CPU to do it). I think it's more important here that the code is clearly understandable and somebody can optimize it once it has been proved out.

But fetching all the node IDs into memory is certainly _ugly_. For Khan Academy, for example, that's a really big query and about 800 kB of memory. I'm sure we could do something clever involving the `lft` / `rft` of leaf nodes that are excluded, bubbling up from them to select the correct topics. There's a danger that it could turn into _a lot_ of queries for a problem where we're really just shuffling primary keys around. So while it would involve less duplicate data in memory, I don't think we're guaranteed that it would be more efficient, particularly for the sort of deployment where someone would be using `exportcontent`. Worth exploring in a branch?

Reply:
We do something very similar in our importability annotation https://github.com/learningequality/kolibri/blob/develop/kolibri/core/content/utils/importability_annotation.py#L36 - the main difference here would be instead of doing annotation based on `some`, we would annotate on `all` children being available. Otherwise the code would be nearly identical. We can then descend down the tree in a read query until we have a set of ids that are marked as 'all' available.

The main trick here is that the write and read all happen within a transaction, and then we roll it back at the end.